library/sha512.c

```diff
@@ -43,6 +43,47 @@
 #if defined(__aarch64__)
 #  if defined(MBEDTLS_SHA512_USE_A64_CRYPTO_IF_PRESENT) || \
     defined(MBEDTLS_SHA512_USE_A64_CRYPTO_ONLY)
+/*
+ * Best performance comes from most recent compilers, with intrinsics and -O3.
+ * Must compile with -march=armv8.2-a+sha3, but we can't detect armv8.2-a, and
+ * can't always detect __ARM_FEATURE_SHA512 (notably clang 7-12).
+ *
+ * GCC < 8 won't work at all (lacks the sha512 instructions)
+ * GCC >= 8 uses intrinsics, sets __ARM_FEATURE_SHA512
+ *
+ * Clang < 7 won't work at all (lacks the sha512 instructions)
+ * Clang 7-12 don't have intrinsics (but we work around that with inline
+ *            assembler) or __ARM_FEATURE_SHA512
+ * Clang == 13.0.0 same as clang 12 (only seen on macOS)
+ * Clang >= 13.0.1 has __ARM_FEATURE_SHA512 and intrinsics
+ */
+#    if !defined(__ARM_FEATURE_SHA512)
+       /* Test Clang first, as it defines __GNUC__ */
+#      if defined(__clang__)
+#        if __clang_major__ < 7
+#          error "A more recent Clang is required for MBEDTLS_SHA512_USE_A64_CRYPTO_*"
+#        elif __clang_major__ < 13 || \
+              (__clang_major__ == 13 && __clang_minor__ == 0 && \
+               __clang_patchlevel__ == 0)
+         /* We implement the intrinsics with inline assembler, so don't error */
+#        elif __clang_major__ > 15
+           /* If https://reviews.llvm.org/D131064 got merged, we should
+            * re-consider above version check. */
+#          pragma clang attribute push (__attribute__((target("sha3"))), apply_to=function)
```

Comment:
There is a good reason for requiring CFLAGS to match the `_USE_A64_CRYTPO_` value specified: the instructions aren't available in all CPU. That was the conscious decision (after discussion with the team) when the SHA-2 acceleration work started.

i.e. if you say you want SHA-512 acceleration, you must compile with the right target CPU flags.

Yes, this is not how (for example) OpenSSL does things, but Mbed TLS is a different flavour of library.

Not only does this move the checks from `check_config.h` to `sha???.c`, it also changes the requirement for building the library with the right command-line flags

Reply:
> There is a good reason for requiring CFLAGS to match the _USE_A64_CRYTPO_ value specified: the instructions aren't available in all CPU. That was the conscious decision (after discussion with the team) when the SHA-2 acceleration work started.

Without `runtime cpu feature detection`, I agree. 
Do you mean `build for target only` ? I am confuse on that. That conflicts with the code.

> Yes, this is not how (for example) OpenSSL does things, but Mbed TLS is a different flavour of library.

At this point, I do not agree. The requirement for AES includes `runtime detection` ( https://github.com/Mbed-TLS/mbedtls/issues/5387#issuecomment-1373523124 ) 

SHA256, SHA512, AESNI and PADLOCK include runtime detection code. So I guess it is in feature list.

> Not only does this move the checks from check_config.h to sha???.c, it also changes the requirement for building the library with the right command-line flags

Yes, it change that. That's the reason why I post this PR to discuss that. And I think that is cheap solution for `illegal instruction` and `Arm64 host build fail`